src/main/scala/org/broadinstitute/dsde/workbench/sam/service/UserService.scala

```diff
@@ -21,15 +21,16 @@ import scala.util.matching.Regex
 /**
   * Created by dvoet on 7/14/17.
   */
-class UserService(val directoryDAO: DirectoryDAO, val cloudExtensions: CloudExtensions, val registrationDAO: RegistrationDAO, blockedEmailDomains: Seq[String])(implicit val executionContext: ExecutionContext, contextShift: ContextShift[IO]) extends LazyLogging {
+class UserService(val directoryDAO: DirectoryDAO, val cloudExtensions: CloudExtensions, val registrationDAO: RegistrationDAO, blockedEmailDomains: Seq[String], tosService: TosService)(implicit val executionContext: ExecutionContext, contextShift: ContextShift[IO]) extends LazyLogging {
 
-  def createUser(user: WorkbenchUser, samRequestContext: SamRequestContext): Future[UserStatus] = {
+  def createUser(user: WorkbenchUser, samRequestContext: SamRequestContext, tosEnforcementEnabled: Boolean = false, tosVersion: Int = 0): Future[UserStatus] = {
     for {
       _ <- UserService.validateEmailAddress(user.email, blockedEmailDomains).unsafeToFuture()
       allUsersGroup <- cloudExtensions.getOrCreateAllUsersGroup(directoryDAO, samRequestContext)
       createdUser <- registerUser(user, samRequestContext).unsafeToFuture()
       _ <- enableUserInternal(createdUser, samRequestContext)
       _ <- directoryDAO.addGroupMember(allUsersGroup.id, createdUser.id, samRequestContext).unsafeToFuture()
+      _ <- if (tosEnforcementEnabled) directoryDAO.addGroupMember(tosService.getTosGroup(tosVersion).unsafeRunSync().get.id, createdUser.id, samRequestContext).unsafeToFuture() else Future.successful(IO.pure(true))
```

Comment:
a couple of things:

1) the value returned by the `else` clause only needs to be a `Future.successful(true)`. the value is ultimately discarded but if you assign this line to a variable you'll see that it returns `Any` because the two possible return types are `Future[Boolean]` and `Future[IO[Boolean]]`

2) I think i would prefer to see this split out into it's individual steps:

```
      tosGroup <- if (tosEnforcementEnabled) tosService.getTosGroup(tosVersion).unsafeToFuture() else Future.successful(None)
      _ <- if (tosEnforcementEnabled && tosGroup.nonEmpty) directoryDAO.addGroupMember(tosGroup.get.id, createdUser.id, samRequestContext).unsafeToFuture() else Future.successful(true)
```

there is another benefit: using `.get` on `Option`s is something that i personally find to be a bit dangerous since it can throw a `NoSuchElementException` and then there's no handling in place to deal with that. doing the `nonEmpty` check at least grants some reassurance that there will always be a ToS group returned. testing for emptiness is logically safe but pattern matching or `getOrElse` are usually preferred, so there is another way to write this that takes it a step further and protects against these exception situations:

```
      tosGroupOpt <- if (tosEnforcementEnabled) tosService.getTosGroup(tosVersion).unsafeToFuture() else Future.successful(None)
      _ <- (tosEnforcementEnabled, tosGroupOpt) match {
        case (true, Some(tosGroup)) => directoryDAO.addGroupMember(tosGroup.id, createdUser.id, samRequestContext).unsafeToFuture()
        case _ => Future.successful(true)
      }
```

disclaimer: both of these compile but i haven't run unit tests or anything. but i _think_ they should both work

i personally prefer the 2nd option but i understand that it's not quite as readable, so i'll leave it up to you!

Reply:
Thanks, Matt! I've updated the code to use the 2nd option.